perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -991,81 +1022,35 @@ public static boolean hasRole(PerunPrincipal perunPrincipal, String role) {
-	 * For role "PERUNADMIN" ignore complementary object.
 	 *
 	 * @param sess                perun session
 	 * @param user                the user for setting role
 	 * @param role                role of user in a session ( PERUNADMIN | VOADMIN | GROUPADMIN | SELF | FACILITYADMIN | VOOBSERVER | TOPGROUPCREATOR | SECURITYADMIN | RESOURCESELFSERVICE | RESOURCEADMIN )
 	 * @param complementaryObject object for which role will be set
 	 */
-	public static void setRole(PerunSession sess, User user, PerunBean complementaryObject, String role) throws AlreadyAdminException {
-		List<PerunBean> complementaryObjects = new ArrayList<>();
-		complementaryObjects.add(complementaryObject);
-		AuthzResolverBlImpl.setRole(sess, user, role, complementaryObjects);
-	}
+	public static void setRole(PerunSession sess, User user, PerunBean complementaryObject, String role) throws AlreadyAdminException, RoleCannotBeManagedException {
+		if (!objectAndRoleManageableByEntity(user, complementaryObject, role)) {
+			throw new RoleCannotBeManagedException("Combination of Role: "+ role +", Object: "+ complementaryObject +" and Entity: "+ user +" cannot be managed.");
+		}
 
-	/**
-	 * Set role for auhtorizedGroup and <b>all</b> complementary objects.
-	 * <p>
-	 * If some complementary object is wrong for the role, throw an exception.
-	 * For role "PERUNADMIN" ignore complementary objects.
-	 *
-	 * @param sess                 perun session
-	 * @param authorizedGroup      the group for setting role
-	 * @param role                 role of user in a session ( PERUNADMIN | VOADMIN | GROUPADMIN | SELF | FACILITYADMIN | VOOBSERVER | TOPGROUPCREATOR | RESOURCESELFSERVICE | RESOURCEADMIN )
-	 * @param complementaryObjects objects for which role will be set
-	 */
-	public static void setRole(PerunSession sess, Group authorizedGroup, String role, List<PerunBean> complementaryObjects) throws AlreadyAdminException {
-		if (complementaryObjects == null || complementaryObjects.isEmpty()) {
-			try {
-				manageRole(sess, SET_ROLE, authorizedGroup, null, role, null);
-				//These exceptions should never happen
-			} catch (GroupNotAdminException | UserNotAdminException ex) {
-				throw new InternalErrorException(ex);
-			}
-		} else {
-			for (PerunBean compObject : complementaryObjects) {
-				try {
-					manageRole(sess, SET_ROLE, authorizedGroup, null, role, compObject);
-					//These exceptions should never happen
-				} catch (GroupNotAdminException | UserNotAdminException ex) {
-					throw new InternalErrorException(ex);
-				}
+		Map<String, Integer> mappingOfValues = createMappingOfValues(user, complementaryObject, role);
+
+		try {
+			authzResolverImpl.setRole(sess, mappingOfValues);
+		} catch (RoleAlreadySetException e) {
+			throw new AlreadyAdminException("User id=" + user.getId() + " is already "+role+" in " + complementaryObject, e);
+		}
+
+		getPerunBl().getAuditer().log(sess, new RoleSetForUser(complementaryObject, user, role));
```

Comment:
With these new generic messages, some other changes must be made. Some components might be dependent on the old format. Unfortunately, I don't know every place where this is used.

Reply:
One workaround would be to log the new message together with the old one (if there was any) for specific roles being set. I guess it depends on how many changes must be done because of this new role management.

Reply:
We have searched for occurrences and it seems that only notifications use "set" event. So in order to use the new one this SQL command have to be executed.
`update pn_regex set regex = 'Role PERUNADMIN was set for.*' where regex = 'User:.*was promoted to PERUNADMIN.*';`